File: ios/Sources/GutenbergKit/Sources/Model/HTTP/EditorHTTPHeaders.swift
import Foundation

/// A case-insensitive collection of HTTP headers.
///
/// This struct provides a type-safe way to work with HTTP headers, with case-insensitive
/// key lookups as required by the HTTP specification. Headers can be created using
/// dictionary literal syntax and are `Codable` for serialization.
///
/// Example usage:
/// ```swift
/// var headers: EditorHTTPHeaders = ["Content-Type": "application/json"]
/// headers["accept"] = "text/html"  // Case-insensitive
/// print(headers["CONTENT-TYPE"])   // "application/json"
/// ```
public struct EditorHTTPHeaders: Sendable, Codable, Equatable, Hashable, ExpressibleByDictionaryLiteral {
    private var elements: [String: String]

    /// Creates headers from a string dictionary.
    ///
    /// - Parameter elements: A dictionary of header names to values.
    init(_ elements: [String: String] = [:]) {
        self.elements = elements
    }

    /// Creates headers from an `HTTPURLResponse`'s `allHeaderFields` dictionary.
    ///
    /// Non-string keys and values are skipped.
    ///
    /// - Parameter elements: A dictionary typically from `HTTPURLResponse.allHeaderFields`.
    init(_ elements: [AnyHashable: Any]) {
        self.elements = [:]

        for (key, value) in elements {
            guard let keyString = key as? String, let valueString = value as? String else {
                continue
            }

            self.elements[keyString] = valueString
        }
    }

    /// Creates headers from a dictionary literal.
    ///
    /// - Parameter elements: Key-value pairs of header names and values.
    public init(dictionaryLiteral elements: (String, String)...) {
        self.elements = elements.reduce(into: [String: String](), { $0[$1.0] = $1.1 })
    }

    /// Accesses the header value for the given key using case-insensitive matching.
    ///
    /// When getting a value, the key is matched case-insensitively against stored headers.
    /// When setting a value, the new key will replace the old key case-insensitively.
    ///
    /// - Parameter key: The header name to look up or set.
    /// - Returns: The header value, or `nil` if no matching header exists.
    subscript(key: String) -> String? {
        get {
            let lowercasedKey = key.lowercased()
            return elements.first { $0.key.lowercased() == lowercasedKey }?.value
        }
        set {
            let lowercasedKey = key.lowercased()
            // Remove any existing key with same case-insensitive match
            if let existingKey = elements.keys.first(where: { $0.lowercased() == lowercasedKey }) {
                elements.removeValue(forKey: existingKey)
            }
            if let newValue {
                elements[key] = newValue
            }
        }
    }

    /// Compares two header collections for equality using case-insensitive key matching.
    public static func == (lhs: EditorHTTPHeaders, rhs: EditorHTTPHeaders) -> Bool {
        guard lhs.elements.count == rhs.elements.count else { return false }
        for (key, value) in lhs.elements {
            guard rhs[key] == value else { return false }
        }
        return true
    }

    /// Returns the headers as a standard string dictionary.
    ///
    /// Useful for passing headers to `URLRequest` or `HTTPURLResponse` APIs.
    public var dictionaryValue: [String: String] {
        self.elements
    }

    func toJSON() throws -> JSON {
        .object(elements.reduce(into: [String: JSON]()) { $0[$1.key] = .string($1.value) })
    }

    /// Returns a new headers collection containing only the specified keys.
    ///
    /// Key matching is case-insensitive for filtering.
    ///
    /// - Parameter keys: The header names to include.
    /// - Returns: A new `EditorHTTPHeaders` containing only the matching headers.
    public func filtering(keys: String...) -> EditorHTTPHeaders {
        let lowercasedKeys = keys.map { $0.lowercased() }
        return EditorHTTPHeaders(self.elements.filter { lowercasedKeys.contains($0.key.lowercased()) })
    }
}

File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/model/http/EditorHTTPHeaders.kt
package org.wordpress.gutenberg.model.http

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * A case-insensitive collection of HTTP headers.
 *
 * This class provides a type-safe way to work with HTTP headers, with case-insensitive
 * key lookups as required by the HTTP specification. Headers are `Serializable` for JSON encoding.
 *
 * Example usage:
 * ```kotlin
 * val headers = EditorHTTPHeaders(mapOf("Content-Type" to "application/json"))
 * headers["accept"] = "text/html"  // Case-insensitive
 * println(headers["CONTENT-TYPE"])   // "application/json"
 * ```
 */
@Serializable(with = EditorHTTPHeadersSerializer::class)
class EditorHTTPHeaders(
    elements: Map<String, String> = emptyMap()
) {
    private val elements: MutableMap<String, String> = elements.toMutableMap()

    /**
     * Accesses the header value for the given key using case-insensitive matching.
     *
     * When getting a value, the key is matched case-insensitively against stored headers.
     * When setting a value, the new key will replace the old key case-insensitively.
     *
     * @param key The header name to look up or set.
     * @return The header value, or `null` if no matching header exists.
     */
    operator fun get(key: String): String? {
        val lowercasedKey = key.lowercase()
        return elements.entries.firstOrNull { it.key.lowercase() == lowercasedKey }?.value
    }

    operator fun set(key: String, value: String?) {
        val lowercasedKey = key.lowercase()
        // Remove any existing key with same case-insensitive match
        val existingKey = elements.keys.firstOrNull { it.lowercase() == lowercasedKey }
        if (existingKey != null) {
            elements.remove(existingKey)
        }
        if (value != null) {
            elements[key] = value
        }
    }

    /**
     * Returns the headers as a standard string map.
     *
     * Useful for passing headers to HTTP request APIs.
     */
    val dictionaryValue: Map<String, String>
        get() = elements.toMap()

    fun toJsonElement(): JsonElement {
        return JsonObject(elements.mapValues { JsonPrimitive(it.value) })
    }

    /**
     * Returns a new headers collection containing only the specified keys.
     *
     * Key matching is case-insensitive for filtering.
     *
     * @param keys The header names to include.
     * @return A new `EditorHTTPHeaders` containing only the matching headers.
     */
    fun filtering(vararg keys: String): EditorHTTPHeaders {
        val lowercasedKeys = keys.map { it.lowercase() }
        val filtered = elements.filter { lowercasedKeys.contains(it.key.lowercase()) }
        return EditorHTTPHeaders(filtered)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is EditorHTTPHeaders) return false
        if (elements.size != other.elements.size) return false
        for ((key, value) in elements) {
            if (other[key] != value) return false
        }
        return true
    }

    override fun hashCode(): Int {
        // Use lowercase keys for consistent hashing
        return elements.entries.sumOf {
            it.key.lowercase().hashCode() + it.value.hashCode()
        }
    }

    companion object {
        val empty = EditorHTTPHeaders()
    }
}

object EditorHTTPHeadersSerializer : KSerializer<EditorHTTPHeaders> {
    private val delegateSerializer = MapSerializer(String.serializer(), String.serializer())

    override val descriptor: SerialDescriptor = delegateSerializer.descriptor

    override fun serialize(encoder: Encoder, value: EditorHTTPHeaders) {
        encoder.encodeSerializableValue(delegateSerializer, value.dictionaryValue)
    }

    override fun deserialize(decoder: Decoder): EditorHTTPHeaders {
        val map = decoder.decodeSerializableValue(delegateSerializer)
        return EditorHTTPHeaders(map)
    }
}
